src/Nethermind/Nethermind.Facade/Eth/RpcTransaction/TransactionForRpc.cs

```diff
@@ -0,0 +1,154 @@
+/// <para><see cref="TransactionForRpc"/> -> JSON (Derived by <c>System.Text.JSON</c> using the runtime type)</para>
+/// </remarks>
+[JsonConverter(typeof(JsonConverter))]
+public abstract class TransactionForRpc
+{
+    [JsonIgnore(Condition = JsonIgnoreCondition.Never)]
+    public virtual TxType? Type => null;
+
+    [JsonIgnore(Condition = JsonIgnoreCondition.Never)]
+    public Hash256? Hash { get; set; }
+
+    [JsonIgnore(Condition = JsonIgnoreCondition.Never)]
+    public long? TransactionIndex { get; set; }
+
+    [JsonIgnore(Condition = JsonIgnoreCondition.Never)]
+    public Hash256? BlockHash { get; set; }
+
+    [JsonIgnore(Condition = JsonIgnoreCondition.Never)]
+    public long? BlockNumber { get; set; }
+
+    [JsonConstructor]
+    public TransactionForRpc() { }
+
+    public TransactionForRpc(Transaction transaction, int? txIndex = null, Hash256? blockHash = null, long? blockNumber = null)
+    {
+        Hash = transaction.Hash;
+        TransactionIndex = txIndex;
+        BlockHash = blockHash;
+        BlockNumber = blockNumber;
+    }
+
+    public virtual Transaction ToTransaction()
+    {
+        return new Transaction
+        {
+            Type = Type ?? default,
+        };
+    }
+
+    public abstract void EnsureDefaults(long? gasCap);
+
+    public class JsonConverter : JsonConverter<TransactionForRpc>
+    {
+        private readonly Type[] _transactionTypes = new Type[Transaction.MaxTxType + 1];
+
+        // TODO: Refactoring transition code
+        public JsonConverter()
+        {
+            RegisterTransactionType(TxType.Legacy, typeof(LegacyTransactionForRpc));
+            RegisterTransactionType(TxType.AccessList, typeof(AccessListTransactionForRpc));
+            RegisterTransactionType(TxType.EIP1559, typeof(EIP1559TransactionForRpc));
+            RegisterTransactionType(TxType.Blob, typeof(BlobTransactionForRpc));
+            // TODO: Add Optimism:
+            // RegisterTransactionType(TxType.DepositTx, typeof(RpcOptimismTransaction))
+        }
+
+        public JsonConverter RegisterTransactionType(TxType type, Type @class)
+        {
+            if (!@class.IsSubclassOf(typeof(TransactionForRpc)))
+            {
+                throw new ArgumentException($"{@class.FullName} is not a subclass of ${nameof(TransactionForRpc)}");
+            }
+
+            _transactionTypes[(byte)type] = @class;
+            return this;
+        }
+
+        public override TransactionForRpc? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            using var document = JsonDocument.ParseValue(ref reader);
+
+            // TODO: For some reason we might get an object wrapped in a String
+            using JsonDocument jsonObject = document.RootElement.ValueKind == JsonValueKind.String
+                ? JsonDocument.Parse(document.RootElement.GetString()!)
+                : document;
```

Comment:
@benaadams Do you know why on RPC parameter execution we might get JSON objects wrapped in strings? For example, instead of getting:

```json
{ "foo": 42 }
```

I get:

```json
"{\"foo\": 42 }"
```

This workaround deserializes the object again so we can actually treat as an object.

Reply:
`JsonDocument` might not be the most efficient API to use. See https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/converters-how-to?pivots=dotnet-8-0#support-polymorphic-deserialization for more details.

Reply:
Yeah not the biggest fan of that, can we try using type discriminators: https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/polymorphism?pivots=dotnet-8-0#customize-the-type-discriminator-name

```
[JsonPolymorphic(TypeDiscriminatorPropertyName = "TxType")]
[JsonDerivedType(typeof(LegacyTransactionForRpc), typeDiscriminator: "Legacy")] 
```

or `[JsonDerivedType(typeof(LegacyTransactionForRpc), typeDiscriminator: 0)]`

as we would have to make it pluginable maybe we should create custom resolver: https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/polymorphism?pivots=dotnet-8-0#configure-polymorphism-with-the-contract-model which would register all the tx types.

Reply:
Here is where I'm currently at when it comes to custom type resolvers:

I've created the following modifier function:

```C#
public static Action<JsonTypeInfo> JsonTypeInfoModifier(params (Type, TxType)[] txTypes)
{
    return jsonTypeInfo =>
    {
        Type baseValueObjectType = typeof(TransactionForRpc);
        if (jsonTypeInfo.Type == baseValueObjectType)
        {
            jsonTypeInfo.PolymorphismOptions = new JsonPolymorphismOptions
            {
                TypeDiscriminatorPropertyName = "type",
                IgnoreUnrecognizedTypeDiscriminators = true,
                UnknownDerivedTypeHandling = JsonUnknownDerivedTypeHandling.FailSerialization,
            };

            foreach (var (type, txType) in txTypes)
            {
                jsonTypeInfo.PolymorphismOptions.DerivedTypes.Add(new JsonDerivedType(type, (byte)txType));
            }
        }
    };
}
```

Our `EthereumJsonSerializer` needs to be modified to support injecting this modifier function, which we can do somewhat similar to how we're currently adding converters:

https://github.com/NethermindEth/nethermind/blob/f825abde15a1db44de82a9d8bda9d2ae7065c2d7/src/Nethermind/Nethermind.Serialization.Json/EthereumJsonSerializer.cs#L99-L106

We add

```C#
private static readonly List<Action<JsonTypeInfo>> _typeInfoResolverModifiers = new();
public static void AddTypeInfoResolverModifier(Action<JsonTypeInfo> modifier)
{
    _typeInfoResolverModifiers.Add(modifier);
        
    JsonOptions = CreateOptions(indented: false);
    JsonOptionsIndented = CreateOptions(indented: true);
}
```

And we modify

https://github.com/NethermindEth/nethermind/blob/f825abde15a1db44de82a9d8bda9d2ae7065c2d7/src/Nethermind/Nethermind.Serialization.Json/EthereumJsonSerializer.cs#L57-L61

to

```C#
private static JsonSerializerOptions CreateOptions(bool indented, IEnumerable<JsonConverter> converters = null, int maxDepth = 64)
{
    IJsonTypeInfoResolver typeInfoResolver = new DefaultJsonTypeInfoResolver();
    foreach (Action<JsonTypeInfo> resolverModifier in _typeInfoResolverModifiers)
    {
        typeInfoResolver = typeInfoResolver.WithAddedModifier(resolverModifier);
    }
        
    var options = new JsonSerializerOptions
    {
        TypeInfoResolver = typeInfoResolver,
        WriteIndented = indented,
```

There are some issues with this approach:

- It does not provide a mechanism to default to a specific type when the type discriminator is missing, instead it defaults to the base class of the hierarchy which in our case is abstract, thus throwing an exception. For our use case we want to default to `LegacyTransactionForRpc` when the discriminator is missing.
- My attempts at getting it to work don't trigger this error (which I get in a toy project that I'm using for testing), instead it fails with a "no available parameterless constructor" exception (note that we do have parameterless constructors). The full exception message is the following:

```Deserialization of types without a parameterless constructor, a singular parameterized constructor, or a parameterized constructor annotated with 'JsonConstructorAttribute' is not supported. Type 'Nethermind.Facade.Eth.RpcTransaction.TransactionForRpc'```